### PR TITLE
Synthetic wavelength calibration data

### PR DESCRIPTION
**Pull request by tepickering**
This PR adds a utility for creating realistic models of 2D spectroscopic images of arc lamp (or sky background) calibration observations. This will be key for testing the various stages of the wavelength calibration code and process as it's developed. Tilts/curvature along the spatial axis is modeled using 1D polynomials. Non-linear dispersion relations are modeled using the ideal disperser model implemented within the FITS WCS standard. Examples of the use of this routine are included in its docstrings. 

A few other notes:

- Use of the `specreduce_data` dependency has been deprecated. Cleaner to just always download/cache the data as needed. 
- https://github.com/pypeit/PypeIt is used as the source for comparison line lists because it has a large selection of well-vetted ones. 
- In the example where a non-linear dispersion relation is created using the WCS ideal disperser model, I create the model by putting together the appropriate FITS header manually and then loading that into `astropy.wcs.WCS`. It would be useful to have a documented functional interface for building WCS spectrograph models that provides meaningful parameter names and units handling. However, I think that is outside of the scope of this PR. It's also an open question whether such a utility should live here or within `specutils`.

### CODE REVIEW

File: specreduce/utils/synth_data.py
Comment (kecnry) [inline specreduce/utils/synth_data.py:117]:
```suggestion
        If ``wcs`` is not provided, this defines the wavelength units of the
```

File: specreduce/utils/synth_data.py
Comment (kecnry) [inline specreduce/utils/synth_data.py:126]:
```suggestion
        Scale factor to apply to amplitudes provided in the linelists
```

File: specreduce/utils/synth_data.py
Comment (kecnry) [inline specreduce/utils/synth_data.py:116]: should this be a `Quantity` or `Unit` object?

File: specreduce/utils/synth_data.py
Comment (kecnry) [inline specreduce/utils/synth_data.py:250]: do we need to check if the unit is valid (`u.Unit` and spectral)?

File: specreduce/utils/synth_data.py
Comment (kecnry) [inline specreduce/utils/synth_data.py:115]: having this between the two arguments that are only applicable if `wcs` is not provided might be confusing - could we move it below `extent` and `wave_unit`?

**Review by kecnry — APPROVED** [review]
Just a few minor comments/suggestions, but otherwise the code and tests look reasonable to me.  This will definitely be useful to use as we write more tests for wavelength calibration and spectral extraction.

File: specreduce/utils/synth_data.py
Comment (tepickering) [reply specreduce/utils/synth_data.py:17]: how about renaming `make_2dspec_image` to `make_2d_trace_image`? that's a more accurate description of what it does. this would open up the possibility of a `make_2d_spec_image` that combines `make_2d_trace_image` and `make_2d_arc_image`. might be worth implementing that as part of this PR.

File: specreduce/utils/synth_data.py
Comment (tepickering) [reply specreduce/utils/synth_data.py:117]: gah, fat fingers...

File: specreduce/utils/synth_data.py
Comment (tepickering) [reply specreduce/utils/synth_data.py:116]: ah ha! good point. it should in fact be `Unit`...

File: specreduce/utils/synth_data.py
Comment (tepickering) [reply specreduce/utils/synth_data.py:115]: i'll at least do that if there's not a more clever way around it...

**Comment by tepickering** [comment]
thanks for the review, @kecnry! i think i addressed all of the comments. i also added a test i forgot to earlier and added a few more checks to flesh out coverage. if there are no other comments/concerns, i'll go ahead and merge in the next day or two.

File: specreduce/calibration_data.py
Comment (pllim) [inline specreduce/calibration_data.py:93]: Instead of rolling your own here, why not use the stuff in `astropy.utils.data` but use your own [conf.dataurl](https://github.com/astropy/astropy/blob/b40081ff722c5ca0a4ff1578abe4587e44c0eeba/astropy/utils/data.py#L99)?

https://github.com/astropy/astropy/blob/main/astropy/utils/data.py

File: specreduce/tests/test_get_reference_file_path.py
Comment (pllim) [inline specreduce/tests/test_get_reference_file_path.py:13]: If your test downloads data, you might wanna mark it with `@pytest.mark.remote_data`.

File: specreduce/calibration_data.py
Comment (tepickering) [reply specreduce/calibration_data.py:102]: the use of the package is deprecated. the repository is still there and will be there unless and until something better comes along.

Comment (pllim) [reply specreduce/calibration_data.py:102]: OK, I opened https://github.com/astropy/specreduce-data/issues/13

File: specreduce/tests/test_get_reference_file_path.py
Comment (tepickering) [reply specreduce/tests/test_get_reference_file_path.py:13]: good call. i'll add that.

File: specreduce/calibration_data.py
Comment (tepickering) [reply specreduce/calibration_data.py:93]: when this was originally implemented years ago, i followed examples in https://docs.astropy.org/en/stable/utils/data.html and just wrapped `astropy.utils.data.download_file()` with some error handling. it looks like those examples haven't changed much since. i'll look into refactoring this to use pure-`astropy` at a higher level. that's out of scope for this PR, though.

Comment (tepickering) [reply specreduce/calibration_data.py:93]: i opened #169 to cover this. it's a little unclear to me how best to handle multiple data URLs with the `conf.dataurl` scheme. will play around more when i get the chance.

File: specreduce/tests/test_get_reference_file_path.py
Comment (tepickering) [reply specreduce/tests/test_get_reference_file_path.py:17]: done. i tweaked the CI workflow so that `--remote-data` is only enabled for the one test where coverage is measured. this should hopefully help with the occasional failures due to downloads timing out. no longer parallel tests getting the same files at the same time...

**Comment by pllim** [comment]
Mine was just a drive-by review and not thorough. I'll leave it to specreduce maintainers to do the rest. Thanks!

**Comment by tepickering** [comment]
drive-by or not, it was useful so thanks again, @pllim!

**Pull request merged**
